dev/src/index.ts

```diff
@@ -1265,19 +1281,95 @@ export class Firestore implements firestore.Firestore {
 
     const writer = bulkWriter ?? this.getBulkWriter();
     writer._verifyNotClosed();
-    const docStream = this._getAllDescendants(
-      ref instanceof CollectionReference
-        ? (ref as CollectionReference<unknown>)
-        : (ref as DocumentReference<unknown>)
-    );
+
+    // A deferred promise that resolves when the recursive delete operation
+    // is completed.
     const resultDeferred = new Deferred<void>();
+
     let errorCount = 0;
     let lastError: GoogleError | BulkWriterError | undefined;
     const incrementErrorCount = (err: Error): void => {
       errorCount++;
       lastError = err;
     };
-    const onStreamEnd = (): void => {
+
+    // Whether all descendants for this query have been fetched.
+    let allDescendantsFetched = false;
+
+    // Whether a query stream is currently in progress. Only one stream can be
+    // run at a time.
+    let streamInProgress = false;
+
+    // The last document snapshot returned by the stream. Used to set the
+    // startAfter() field in the subsequent stream.
+    let lastDocumentSnap: QueryDocumentSnapshot;
+
+    // The number of pending BulkWriter operations. Used to determine when the
+    // next query can be run.
+    let pendingPromiseCount = 0;
```

Comment:
```suggestion
    let pendingOpsCount = 0;
```

It it is not clear that the limit is the number of outstanding promises, so I suggest using `ops` instead.

Reply:
done.